src/fanControl.ino

```diff
@@ -0,0 +1,279 @@
+//Adafruit_PWMServoDriver pwm = Adafruit_PWMServoDriver(0x40, Wire);
+
+/*
+Command structure:
+Byte 3:  Command ID
+Byte 2:  parameter 1
+Byte 1:  Parameter 2
+Byte 0:  Response, command response status
+
+The Arduino responds to every command with 4 bytes, Command ID, parameter 1 & 2 and CMD_OK/COMMAND_ERROR
+*/
+
+#define FAN_ID 5
+
+#define CMD_UNKNOWN 0x00  //  Unknown command received
+#define CMD_SPEED  0x01   //  Set duty cycle/pulse width
+#define CMD_RELAY  0x02   //  Set/Reset the relay
+#define CMD_FAN_ID 0x03   //  Send a custom ID number.
+#define CMD_ERROR  0xFF   //  Error in command parameters
+#define CMD_OK 0x01       //  Command was successfully handled
+
+#define PIN_RELAY_1 0x08
+#define PIN_RELAY_2 0x09
+#define PIN_RELAY_3 0x0A
+#define PIN_RELAY_4 0x0B
+#define PIN_RELAY_5 0x0C
+#define PIN_RELAY_6 0x0D
+
+
+#define RELAY_SET 0x01
+#define RELAY_RESET 0x00
+
+uint8_t inBuf[4];
+uint8_t outBuf[4];
+uint8_t channel = 0;  // PWM channel id
+uint8_t duty = 0;     // Duty cycle for the fan.
+
+//uint32_t cmd, response;
+
+void setup() 
+{
+  pinMode(LED_BUILTIN, OUTPUT); // initialize digital pin LED_BUILTIN as an output.
+  digitalWrite(LED_BUILTIN, LOW);    // turn the LED off
+  pinMode(PIN_RELAY_1, OUTPUT);   
+  pinMode(PIN_RELAY_2, OUTPUT);   
+  pinMode(PIN_RELAY_3, OUTPUT);    
+  pinMode(PIN_RELAY_4, OUTPUT);    
+  pinMode(PIN_RELAY_5, OUTPUT);    
+  pinMode(PIN_RELAY_6, OUTPUT);    
+  Serial.begin(115200);
+  //Serial.println("PWM 0 ready");
+  pwm.begin();
+  /*
+   * In theory the internal oscillator (clock) is 25MHz but it really isn't
+   * that precise. You can 'calibrate' this by tweaking this number until
+   * you get the PWM update frequency you're expecting!
+   * The int.osc. for the PCA9685 chip is a range between about 23-27MHz and
+   * is used for calculating things like writeMicroseconds()
+   * Analog servos run at ~50 Hz updates, It is importaint to use an
+   * oscilloscope in setting the int.osc frequency for the I2C PCA9685 chip.
+   * 1) Attach the oscilloscope to one of the PWM signal pins and ground on
+   *    the I2C PCA9685 chip you are setting the value for.
+   * 2) Adjust setOscillatorFrequency() until the PWM update frequency is the
+   *    expected value (50Hz for most ESCs)
+   * Setting the value here is specific to each individual I2C PCA9685 chip and
+   * affects the calculations for the PWM update frequency. 
+   * Failure to correctly set the int.osc value will cause unexpected PWM results
+   */
+  pwm.setOscillatorFrequency(27000000);
+  pwm.setPWMFreq(1600);  // This is the maximum PWM frequency
+
+  // if you want to really speed stuff up, you can go into 'fast 400khz I2C' mode
+  // some i2c devices dont like this so much so if you're sharing the bus, watch
+  // out for this!
+  Wire.setClock(400000);
+}
+
+void loop() 
+{
+  int i = 0;
+  while (Serial.available() > 0) 
+  {
+    i = i + Serial.readBytes(inBuf, 4);

```

Comment:
This will read 4 bytes and block until the determined 4 bytes have been read. There is absolutely no need for `i` or testing if `i == 4`. I think.

Reply:
I noticed that the Python serial read command had a timeout, so when i manually sent a test response the script only received half the message sometimes before timing out.

Reply:
But a timeout on the python side (the app) would not affect or be affected by this `i` counter. This would only be useful if the python app was to send the 4 bytes in slow batches. But a `ser.write` command of 4 bytes should take a ms or less on a laptop.